### PR TITLE
Hm 2 dev

### PR DESCRIPTION
**Pull request by AndrewPolishchuc**

### CODE REVIEW

File: src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+@Dao
+public class OrderDaoImpl implements OrderDao {
+    @Override
+    public Order completeOrder(ShoppingCart shoppingCart) {
+        Order order = new Order(shoppingCart.getUserId());
+        Storage.addOrder(order);
+        return order;
```
Comment (SetMyDream) [inline src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java]: also you need to clear shopCart up to Checklist

**Review by SetMyDream — CHANGES_REQUESTED** [review]

File: src/main/java/com/internet/shop/dao/OrderDao.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package com.internet.shop.dao;
+
+import com.internet.shop.model.Order;
+import com.internet.shop.model.ShoppingCart;
+import java.util.List;
+import java.util.Optional;
+
+public interface OrderDao {
+    Order completeOrder(ShoppingCart shoppingCart);
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/OrderDao.java]: we should have this method only on the service layer
there we should have only create() method

File: src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java (hunk 1 above)
Comment (AndrewPolishchuc) [reply src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java]: Sure

File: src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+@Dao
+public class OrderDaoImpl implements OrderDao {
+    @Override
+    public Order completeOrder(ShoppingCart shoppingCart) {
+        Order order = new Order(shoppingCart.getUserId());
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java]: all this logic should be at the service layer

File: src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        Storage.addOrder(order);
+        shoppingCart.setProducts(new ArrayList<>());
+        return order;
+    }
+
+    @Override
+    public Optional<Order> get(Long id) {
+        return getAll().stream()
+                .filter(order -> order.getUserId().equals(id))
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java]:
```suggestion
                .filter(order -> order.getId().equals(id))
```

File: src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    @Override
+    public Order create(Order order) {
+        Storage.addOrder(order);
+        return order;
+    }
+
+    @Override
+    public Order update(Order order) {
+        IntStream.range(0, Storage.orders.size())
+                .filter(i -> Storage.orders.get(i).getUserId().equals(order.getUserId()))
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java]:
```suggestion
                .filter(i -> Storage.orders.get(i).getId().equals(order.getId()))
```

File: src/main/java/com/internet/shop/dao/impl/ShoppingCartDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+@Dao
+public class ShoppingCartDaoImpl implements ShoppingCartDao {
+    @Override
+    public ShoppingCart create(ShoppingCart shoppingCart) {
+        Storage.addShoppingCart(shoppingCart);
+        return shoppingCart;
+    }
+
+    @Override
+    public Optional<ShoppingCart> get(Long id) {
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/dao/impl/ShoppingCartDaoImpl.java:20]: should we find shoppingcart by id or user id?

File: src/main/java/com/internet/shop/service/impl/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import com.internet.shop.model.Order;
+import com.internet.shop.model.ShoppingCart;
+import com.internet.shop.service.OrderService;
+import com.internet.shop.service.ShoppingCartService;
+import java.util.List;
+
+@Service
+public class OrderServiceImpl implements OrderService {
+    @Inject
+    OrderDao orderDao;
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/service/impl/OrderServiceImpl.java]: access modifiers?

File: src/main/java/com/internet/shop/service/impl/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    @Override
+    public ShoppingCart create(ShoppingCart shoppingCart) {
+        return shoppingCartDao.create(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCart addProduct(ShoppingCart shoppingCart, Product product) {
+        List<Product> listOfProducts = shoppingCart.getProducts();
+        listOfProducts.add(product);
+        shoppingCart.setProducts(listOfProducts);
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/service/impl/ShoppingCartServiceImpl.java]: how about passing data by the link?

File: src/main/java/com/internet/shop/service/impl/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        return shoppingCartDao.update(shoppingCart);
+    }
+
+    @Override
+    public boolean deleteProduct(ShoppingCart shoppingCart, Product product) {
+        List<Product> products = shoppingCart.getProducts();
+        boolean removedProduct = products.remove(product);
+        shoppingCart.setProducts(products);
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/service/impl/ShoppingCartServiceImpl.java]: the same) do we need this line?

File: src/main/java/com/internet/shop/service/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+import com.internet.shop.lib.Inject;
+import com.internet.shop.lib.Service;
+import com.internet.shop.model.User;
+import com.internet.shop.service.UserService;
+import java.util.List;
+
+@Service
+public class UserServiceImpl implements UserService {
+    @Inject
+    UserDao userDao;
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/service/impl/UserServiceImpl.java]: ?

File: src/main/java/com/internet/shop/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Override
+    public User create(User user) {
+        userDao.create(user);
+        return user;
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/service/impl/UserServiceImpl.java]:
```suggestion
        return userDao.create(user);
```

File: src/main/java/com/internet/shop/dao/OrderDao.java (hunk 2 above)
Comment (AndrewPolishchuc) [reply src/main/java/com/internet/shop/dao/OrderDao.java]: Yes, messed up packages. Done

File: src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java (hunk 3 above)
Comment (AndrewPolishchuc) [reply src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java]: Done

File: src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java (hunk 4 above)
Comment (AndrewPolishchuc) [reply src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java]: Done

File: src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java (hunk 5 above)
Comment (AndrewPolishchuc) [reply src/main/java/com/internet/shop/dao/impl/OrderDaoImpl.java]: Done

File: src/main/java/com/internet/shop/service/impl/OrderServiceImpl.java (hunk 7 above)
Comment (AndrewPolishchuc) [reply src/main/java/com/internet/shop/service/impl/OrderServiceImpl.java]: Sure.

File: src/main/java/com/internet/shop/service/impl/UserServiceImpl.java (hunk 11 above)
Comment (AndrewPolishchuc) [reply src/main/java/com/internet/shop/service/impl/UserServiceImpl.java]: Done

File: src/main/java/com/internet/shop/dao/impl/ShoppingCartDaoImpl.java (hunk 6 above)
Comment (AndrewPolishchuc) [reply src/main/java/com/internet/shop/dao/impl/ShoppingCartDaoImpl.java:20]: We have to find the cart by id. Done.

File: src/main/java/com/internet/shop/service/impl/UserServiceImpl.java (hunk 10 above)
Comment (AndrewPolishchuc) [reply src/main/java/com/internet/shop/service/impl/UserServiceImpl.java]: Added access modifier.

File: src/main/java/com/internet/shop/service/impl/ShoppingCartServiceImpl.java (hunk 8 above)
Comment (AndrewPolishchuc) [reply src/main/java/com/internet/shop/service/impl/ShoppingCartServiceImpl.java]: Without these unnecessary actions. Done

File: src/main/java/com/internet/shop/service/impl/ShoppingCartServiceImpl.java (hunk 9 above)
Comment (AndrewPolishchuc) [reply src/main/java/com/internet/shop/service/impl/ShoppingCartServiceImpl.java]: We don't need them. Done.

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**